### ISSUE TITLE
V7 and V9: CSRF is Missing in HTTP GET/POST Requests 

### ISSUE THREAD

**Issue by thepcn3rd** [PHPFusion/PHPFusion#429]
After testing the php-fusion software I found that you have not implemented CSRF tokens to prevent resubmissions of administration and member items.  This presents multiple issues.  I will leave a couple of examples below.  All testing was done on a local installation of PHP-Fusion 7.02.07.  Also I confirmed https://www.exploit-db.com/exploits/35206/ of SQL Injection.

When logged in as an admin and you change a password for a member of the forum.  This allows the replay of that action until the admin logs in again after this action is performed.  Again the admin would login, his password_salt and password would change in the database, an aid value is create of an MD5 of the new password and truncated to be 16 characters long.  Then the admin changes the members password.  This is captured through an infected computer or in-browser malware.  Then the changes can be replayed to change the password and email to whatever the miscreant would like.

The second action is the creation of a new member by the admin.  If these packets are captured another user can be created as that admin when the packets are replayed.  This action can occur until the admin logs in a second time and the password_salt and password are modified in the database changing the aid that is used to validate the transaction.

If you have further questions please ask.  Thank you.

**Comment by meangczac** [comment]
Hi, can you carry out similar test on the version 9? We're no longer updating version 7.xx as we're approaching a total release of the new version.

Albeit the situation is upon the fact that the administrator has malware in his browser that can capture $_POST packets, and the issue is about a missing token solution to prevent replaying/recreating the whole $_POST event by the malware owner. Is that correct?

**Comment by thepcn3rd** [comment]
Yes I can. I do not have time today though.

**Comment by meangczac** [comment]
No worry there mate. Thanks for all the help. Just give me a holler here when you have results. I'll leave the issue open. Cheers!

**Comment by thepcn3rd** [comment]
I recently tested version 9 and found the same vulnerability.  This vulnerability occurs even if the administrator has logged out due to the aid value not changing until after the administrator logs in again.  This would also occur on all information that is done through a GET or a POST request that someone could replay.  For example:  Capture an add member action in ZAP.  Then copy the POST request into the Manual Request Editor and then change the username, password, or email and then send it.  It will create that user regardless if the admin is logged in or not.

Thank you.

**Comment by thepcn3rd** [comment]
A CSRF token would not allow this action to occur.  The CSRF token would associate with the PHP Session and then change every time information is sent as a POST or a GET.  Then when the response to that CSRF token occurs the token is modified preventing a resend of previous POST or GET information.

This is a good link to follow.  It mentions when you are authenticated.  Your issue would occur regardless if they are authenticated or not authenticated.  hxxps://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet

**Comment by thepcn3rd** [comment]
I like the new layout and design of PHP-Fusion.  Keep up the great work that you guys are doing.  If I find anything else I will open a new issue.

**Comment by meangczac** [comment]
Hi, thank you for doing penetration testing for us. 

<blockquote>
For example: Capture an add member action in ZAP. Then copy the POST request into the Manual Request Editor and then change the username, password, or email and then send it. It will create that user regardless if the admin is logged in or not.
</blockquote>

Mind I ask which file is the POST sent to, as I need to replicate this in my machine. Once I can replicate this in my local, I'll be able to cure the defects.

Thank you.

**Comment by thepcn3rd** [comment]
In your last comment you asked which file is the POST sent to.  This vulnerability would apply to any page that does GET or POST requests that could be captured and then resent.  I have recorded a couple of examples of where I have abused the fact that CSRF tokens are not being used.  These recordings can be found at http://scriptkitty.work/recordings/.  

The recording named "csrf-phpfusionv9.webm" is where after capturing the admin creating a user, then replaying the POST request will create another user.  This is done twice to demonstrate that it can be done after the admin does it and after the admin is logged out (Noting that the cookie is set to 2 days is rather long also).

The second recording called "csrf-phpfusionv9-resetEmail-resetUsername.webm" is where a member requests an administrator to change his email or make a modification to the profile.  Then by replaying the POST request it can be done for that user.  Then from a miscreants point of view can start looking up other users and modifying their email address to be the miscreants email address.

Without CSRF it could be a member posting to a forum or posting a picture and then the traffic replay will post additional links or statements that were not stated or pictures that they did not upload but a miscreant could.

CSRF tokens are very valuable to have on every page that does a POST or GET request that has to do with submitting information, changing or removing information.

**Comment by meangczac** [comment]
Thanks for the video! 

I'll try to replicate now. I have already coded a solution, but yet to be tested. I'll test now and update by today. This is what I have been waiting for :+1: 

Edit. Looks like it's not as simple as it should be, for optimization reasons, I need to test several methods.

**Comment by meangczac** [comment]
Ah, I typed the wrong issue number. The fix e804d84, please check if it solves the issue. Now, posting using the used token will not go through member add or any of the subsystem functions.

**Comment by meangczac** [comment]
iAUTH now is now reinforced by Session-Token

**Comment by meangczac** [comment]
The fix: https://www.youtube.com/watch?v=5eLfA_ZEujQ&feature=youtu.be

I gave credits under your username as I do not know your real name in the vid. There was a flaw in the CSRF method in the logical interpretations of the OWASP mechanics, but I covered the whole thanks to your report. 

We had about 0.6-0.7s clock on each page render, but I am not sure if due to the installation of ZAP software, now the render time seems to clock at 1.0-1.02 ish :/ I hope that's just happening my server. Even if real, I hope the community will not mind as the speed can be reimbursed through OpCache.

Cheers.
